Customize daguerre's directory path

fitodic opened the pull request:
Construct the directory path where the adjusted images will be saved to
using the MD5 hash algorithm.

Can be customized using the **DAGUERRE_PATH** setting set in the project's
settings. If left unspecified, the default value will be used, i.e 'dg'. The
maximum length of the specified string is 13 characters.

Example:
* Default: dg/ce/2b/7014c0bdbedea0e4f4bf.jpeg
* Custom (DAGUERRE_PATH = 'img'): img/ce/2b/7014c0bdbedea0e4f4bf.jpeg

Known issue:
If the extracted hash string is 'ad', ad blockers will block the
image. All occurrences of 'ad' will be replaced with 'ag' since the MD5
hash produces letters from 'a' to 'f'.

Other:
* Improve Unicode compatibility
* Test the *upload_to* function

Connected to issue #102 

A special thanks to @mislavcimpersak

melinath commented on daguerre/models.py:
```diff
@@ -99,17 +109,57 @@ def delete_adjusted_images(sender, **kwargs):
+    if not first_dir or len(first_dir) > 13:
+        msg = ('The DAGUERRE_PATH value is more than 13 characters long! '
+               'Falling back to the default value: "dg".')
+        warnings.warn(msg)
```
This warning is should only be sent if first_dir is actually too long - currently, sent in that case _and_ if it's a falsey value.

melinath commented on daguerre/models.py:
```diff
@@ -99,17 +109,57 @@ def delete_adjusted_images(sender, **kwargs):
+    # https://docs.djangoproject.com/en/dev/_modules/django/contrib/auth/hashers/
+    # https://github.com/django/django/blob/master/django/contrib/auth/hashers.py#L524
+    str_for_hash = force_bytes('{} {}'.format(filename, datetime.utcnow()))
+    hash_for_dir = hashlib.md5(str_for_hash).hexdigest().replace('ad', 'ag')
```
(Just what you said in the PR re: `ad` getting blocked would be great.)

fitodic replied:
I've added a comment, but I've kept the more detailed explanation in the function's docstring. This way it can appear in the IDE's *show documentation* popup.

melinath commented on daguerre/models.py:
```diff
@@ -99,17 +109,57 @@ def delete_adjusted_images(sender, **kwargs):
+      produces letters from 'a' to 'f'.
+    """
+
+    first_dir = None
```
or maybe 
```python
first_dir = getattr(settings, 'DAGUERRE_ADJUSTED_IMAGE_PATH`, None) or DEFAULT_ADJUSTED_IMAGE_PATH
```

to handle the case where the setting is set to a falsey value.

melinath commented on daguerre/tests/unit/test_models.py:
```diff
@@ -50,3 +52,41 @@ def test_delete_adjusted_images__delete(self):
+        self.instance = None
+        self.filename = '7014c0bdbedea0e4f4bf.jpeg'
+
+    def test_upload_to__default_upload_dir(self):
```
Ideally these would use mock to check whether the warning was sent and make sure it only pass if it was / wasn't, as appropriate. But I'd be fine merging without that being added to the tests as long as the behavior is correct.

fitodic replied:
I've modified the tests according to the official Python docs regarding [warnings](https://docs.python.org/2/library/warnings.html#testing-warnings).

fitodic commented:
All your suggestions have been taken into consideration and implemented. I've added a section to the *daguerre's* documentation regarding the newly available custom setting.

melinath commented on docs/guides/settings.rst:
```diff
@@ -0,0 +1,21 @@
+    DAGUERRE_ADJUSTED_IMAGE_PATH = 'img'
+
+
+**WARNING:** The maximum length of the ``DAGUERRE_ADJUSTED_IMAGE_PATH`` string
```
This should use an rst [admonition](http://docutils.sourceforge.net/docs/ref/rst/directives.html#specific-admonitions).

melinath commented on docs/guides/settings.rst:
```diff
@@ -0,0 +1,21 @@
+
+**WARNING:** The maximum length of the ``DAGUERRE_ADJUSTED_IMAGE_PATH`` string
+is 13 characters. If the string has more than 13 characters, it will gracefully
+fall back to the the default value, i.e. ``dg/``
```
This line seems to imply the default value has a slash at the end

fitodic commented:
**DEFAULT_ADJUSTED_IMAGE_PATH**: The name of the local variable was changed according to the convention and to better match the model's name. 

**Documentation**: Along with removing the trailing slash from the default value, the _custom settings_ section was revised and simplified by adding a couple of examples.

melinath commented on daguerre/tests/unit/test_models.py:
```diff
@@ -50,3 +54,63 @@ def test_delete_adjusted_images__delete(self):
         AdjustedImage.objects.get(pk=adjusted1.pk)
+
+
+class AdjustesImageUploadToTestCase(BaseTestCase):
```
Typo - should be Adjuste**d**ImageUploadToTestCase

melinath replied:
Sorry I didn't notice this sooner :-p

fitodic replied:
Nice work, better late than never! :smile: 

The typo is fixed.

fitodic commented:
They typo was fixed and pushed.

fitodic commented:
The **daguerre clean** management command and the tests covering it have been updated to reflect the newly introduced changes to the image path.

melinath commented:
Merged, thanks for your hard work on this!

fitodic commented:
My pleasure, thanks for your help!
